### PR TITLE
Block Hours Consumer

### PR DESCRIPTION
**Pull request by DevAseel**
**Summary of change:**
- Added a new micro-service, block hours consumer, which uses all of the block hour micro-services in one class. 

**List of tasks:** 

- [x] create types.
- [x] create interface.
- [x] create class.
- [x] create example.
- [x] create unit test.

**Steps for testing**
- Make sure your local git is up to date
```bash
git pull
```

- Checkout the branch `feature/block-hours-consumer`
```bash
git checkout feature/block-hours-consumer
```

- Run the test across the test-net
```bash
npm run test
```

- Confirm block-hour-consumer test has passed.
- Run example.

### CODE REVIEW

**Review by specialistvlad — DISMISSED** [review]
LGTM

File: src/services/datastore/BlockHourFinder.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    const repository: IRepository | undefined = this.datastore.getRepository(BlockHour);
+    if (!repository) {
+      const message = 'failed to resolve BlockHour Repository';
+      this.error(message);
+      return { success: false, message };
+    }
+
+    try {
+      // find block entity for given block number
+      const blockHours: BlockHour[] = await repository.find({ order: { id: 'DESC' } });
```
Comment (jjfluid) [inline src/services/datastore/BlockHourFinder.ts:37]: May want to limit the query to 1, like this:

```suggestion
      const blockHours: BlockHour[] = await repository.find({
        order: { id: 'DESC' },
        take: 1
      });
```
Otherwise we may be querying the whole table unnecessarily.
I guess `findOne` or `findOneBy` would also do the trick.

File: src/services/checkers/BlockHourChecker.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  /**
+   * checks if block exist in datastore, and saves new block records.
+   * @param blockTimestamp the timestamp of the desired block
+   * @param latestBlockHourRecord last block hour record in datastore
+   */
+
+  async apply(blockTimestamp: Date, latestBlockHourRecord: Date): Promise<BlockHourCheckerResult> {
+    const nextHour = new Date(latestBlockHourRecord.setHours(latestBlockHourRecord.getHours() + 1));
```
Comment (jjfluid) [inline src/services/checkers/BlockHourChecker.ts:48]: Will this handle the case where `latestBlockHourRecord` is the last hour of the day? Because in that case, it should also increment the date. 

`setHours(24)` does noat actually update the date, just sets the `hours` to `00`

File: src/services/datastore/BlockFinderByTimestamp.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    if (!repository) {
+      const message = 'failed to resolve Block Repository';
+      this.error(message);
+      return { success: false, message };
+    }
+
+    try {
+      // find block entity for given block number
+      const block: Block = await repository.findOneBy({ timestampUtc: LessThanOrEqual(timestamp) });
```
Comment (jjfluid) [inline src/services/datastore/BlockFinderByTimestamp.ts:40]: May need to apply sort order to this with ` order: { timestampUtc: 'DESC' },` to get the latest

File: src/services/datastore/BlockHourFinder.ts (hunk 1 above)
Comment (jjfluid) [reply src/services/datastore/BlockHourFinder.ts:37]: Also, probably want to sort by a column that guarantees the time series. `hourlyTimestampUtc` is a good candidate, but actually any column other than `id` would suffice.

File: src/services/parsers/BlockHoursParser.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      const currentTime = new Date();
+      const hourlyTimestampUtc = currentTime;
+      if (!block.blockNumber || !block.timestampUtc) {
+        const message = 'Failed to parse block, blockNumber or/and timestamp are undefiend';
+        return { success: false, error: true, message };
+      }
+      const fields = {
+        blockNumber: block.blockNumber,
+        blockTimestampUtc: block.timestampUtc,
+        hourlyTimestampUtc: hourlyTimestampUtc,
```
Comment (jjfluid) [inline src/services/parsers/BlockHoursParser.ts:34]: This needs to represent the end of the hour, but it was only assigned the current time here:
```
      const currentTime = new Date();
      const hourlyTimestampUtc = currentTime;
```

Either convert the Date in this parser, or include the hourlyTimestampUtc as a param.

File: src/services/checkers/BlockHourChecker.ts (hunk 2 above)
Comment (DevAseel) [reply src/services/checkers/BlockHourChecker.ts:48]: @jjfluid It actually does, I've tested it. Try changing the time in `blockHourChecker.ts` example, you will see that the date also increments.

File: src/services/datastore/BlockFinderByTimestamp.ts (hunk 3 above)
Comment (DevAseel) [reply src/services/datastore/BlockFinderByTimestamp.ts:40]: I felt we didn't need it because the timestamp filed is a unique filed and the TypeORM findOneBy method returns the latest field by default, however, I changed it to this, just to sure: 

```Typescript
      const block: Block = await repository.find({
        where: { timestampUtc: LessThanOrEqual(timestamp) },
        order: { timestampUtc: 'DESC' },
        take: 1,
      });
```

**Comment by DevAseel** [comment]
@jjfluid Done, please check and let me know. 😁

File: src/services/checkers/BlockHourChecker.ts (hunk 2 above)
Comment (specialistvlad) [reply src/services/checkers/BlockHourChecker.ts]: @DevAseel let's have here test case for this edge scenario
@jjfluid thank you for mentioning that

**Comment by DevAseel** [comment]
- Will add a test case to the unit test where hourlyBlockHourRecord is the last hour of the day.

**Review by specialistvlad — DISMISSED** [review]

File: src/services/checkers/BlockHourChecker.ts (hunk 2 above)
Comment (jjfluid) [reply src/services/checkers/BlockHourChecker.ts]: Ok.  I was just trying [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/setHours) yesterday in the "Try it" box and it didnt seem to increment the date. Today it does...I must have been half asleep when I tried yesterday. 

If we wanted a test case, we could even isolate the logic for preparing the hourly timestamp into its own function?  Maybe overkill, but it would no longer be a question

Comment (DevAseel) [reply src/services/checkers/BlockHourChecker.ts]: haha no worries! 
Yeah I can definitely do that! Will have a class of nextHourCalculator, will that be good?

Comment (jjfluid) [reply src/services/checkers/BlockHourChecker.ts]: just looked at you unit tests. on second thought - just ensure there are edge cases. no need for separate class

File: src/services/datastore/BlockFinderByTimestamp.ts (hunk 3 above)
Comment (DevAseel) [reply src/services/datastore/BlockFinderByTimestamp.ts]: It will just take the latest record created that matches the given condition, at least that's how it has been behaving for me. And yeah, I find TypeORM documentation not comprehensive as well 😓 , it's hard to tell.

File: src/services/checkers/BlockHourChecker.ts (hunk 2 above)
Comment (DevAseel) [reply src/services/checkers/BlockHourChecker.ts]: Okay! I've included one edge case, and I'm planing to re-visit all of the unit tests to make them more robust, I'm sure I've missed some scenarios/cases, please let me know if you find any.

File: src/services/datastore/BlockFinderByTimestamp.ts (hunk 3 above)
Comment (jjfluid) [reply src/services/datastore/BlockFinderByTimestamp.ts]: Ok, so better use a sort condition on the timestamps in the `findOneBy`. If we ever have an issue and need to delete records, or create a missing record, the order of records created will be broken.

File: src/services/checkers/BlockHourChecker.ts (hunk 2 above)
Comment (jjfluid) [reply src/services/checkers/BlockHourChecker.ts]: Yeah I see it. Good

File: src/services/datastore/BlockFinderByTimestamp.ts (hunk 3 above)
Comment (DevAseel) [reply src/services/datastore/BlockFinderByTimestamp.ts]: I'm no longer using `findOneBy`, I'm using `find` and taking one row only:

```typescript
      const block: Block = await repository.find({
        where: { timestampUtc: LessThanOrEqual(timestamp) },
        order: { timestampUtc: 'DESC' },
        take: 1,
      });
```

Do you think it's sufficient?

Comment (DevAseel) [reply src/services/datastore/BlockFinderByTimestamp.ts]: The reason for the above is that I think sorting is not available while using `findOneBy`, and `findOneBy` returns a single object while `find` returns an array of objects.

Comment (jjfluid) [reply src/services/datastore/BlockFinderByTimestamp.ts]: okay sounds good to me

**Review by jjfluid — DISMISSED** [review]

**Comment by DevAseel** [comment]
@jjfluid could you please approve this one again, had to resolve conflicts.

**Review by specialistvlad — APPROVED** [review]

**Review by jjfluid — APPROVED** [review]
LGTM
